packages/visual-editing/src/overlay-components/components/UnionInsertMenuOverlay.tsx

```diff
@@ -0,0 +1,192 @@
+import type {SchemaNode, SchemaUnionNode} from '@repo/visual-editing-helpers'
+import {AddIcon} from '@sanity/icons'
+import type {SchemaType} from '@sanity/types'
+import {Button, Flex} from '@sanity/ui'
+import {
+  useCallback,
+  useRef,
+  useState,
+  type FunctionComponent,
+  type HTMLProps,
+  type MouseEvent,
+} from 'react'
+import styled from 'styled-components'
+import type {ElementNode, OverlayComponent} from '../../types'
+import {useDocuments} from '../../ui/optimistic-state/useDocuments'
+import {getArrayInsertPatches} from '../../util/mutations'
+import {InsertMenuPopover} from './InsertMenu'
+
+const AddButton = styled(Button)`
+  position: relative;
+  transform: var(--add-button-position);
+
+  --add-button-position: translateY(0);
+  [data-position='top'] & {
+    --add-button-position: translateY(-50%);
+  }
+  [data-position='right'] & {
+    --add-button-position: translateX(50%);
+  }
+  [data-position='bottom'] & {
+    --add-button-position: translateY(50%);
+  }
+  [data-position='left'] & {
+    --add-button-position: translateX(-50%);
+  }
+`
+const HoverAreaRoot = styled(Flex)`
+  pointer-events: all;
+  height: var(--hover-area-height);
+  width: var(--hover-area-width);
+
+  --hover-area-height: 100%;
+  --hover-area-width: 100%;
+  &[data-position='top'],
+  &[data-position='bottom'] {
+    --hover-area-height: 48px;
+  }
+  &[data-position='right'],
+  &[data-position='left'] {
+    --hover-area-width: 48px;
+  }
+`
+
+const HoverArea: FunctionComponent<{
+  element: ElementNode
+  hoverAreaExtent: HTMLProps<HTMLDivElement>['height' | 'width']
+  node: SchemaUnionNode
+  onAddUnion: (insertPosition: 'before' | 'after', name: string) => void
+  position: 'top' | 'right' | 'bottom' | 'left'
+}> = (props) => {
+  const {element, hoverAreaExtent, node, onAddUnion, position} = props
+  const [showButton, setShowButton] = useState(false)
+  const onEnter = useCallback(() => {
+    setShowButton(true)
+  }, [])
+  const onLeave = useCallback(() => {
+    setShowButton(false)
+  }, [])
+  const ref = useRef<HTMLDivElement | null>(null)
+
+  const bubbleEvent = useCallback(
```

Comment:
I think `bubble` may be slightly confusing here as the event isn't bubbling in the traditional sense. `transfer` or `delegate` or something cuold be good